traits/trait_dict_object.py

```diff
@@ -62,8 +56,299 @@ def __init__(self, added=None, changed=None, removed=None):
+
+        This simply calls all notifiers provided by the class, if any.
+        The notifiers are expected to have the signature::
+
+            notifier(trait_dict, added, changed, removed)
+
+        Any return values are ignored.
+
+        """
+
+        for notifier in self.notifiers:
+            notifier(self, added, changed, removed)
+
+    # -- dict interface -------------------------------------------------------
+
+    def __setitem__(self, key, value):
+        """ Set a value for the key, implements self[key] = value.
+
+        Parameters
+        ----------
+        key : A hashable type.
+            The key for the value.
+        value : any
+            The value to set for the corresponding key.
+
+        """
+
+        removed = {}
+        validated_key = self.key_validator(key)
+        validated_value = self.value_validator(value)
+
+        if key in self:
+            changed = {key: self[key]}
+            added = {}
+        else:
+            changed = {}
+            added = {validated_key: validated_value}
+
+        super().__setitem__(validated_key, validated_value)
+        self.notify(added, changed, removed)
+
+    def __delitem__(self, key):
+        """ Delete the item from the dict indicated by the key.
+
+        Parameters
+        ----------
+        key : A hashable type.
+            The key to be deleted.
+
+        Raises
+        ------
+        KeyError
+            If the key is not found.
+
+
+        """
+        removed = {key: self[key]}
+        super().__delitem__(key)
+        self.notify(added={}, changed={}, removed=removed)
+
+    def clear(self):
+        """ Remove all items from the dict. """
+        if self != {}:
+            removed = self.copy()
+            super().clear()
+            self.notify(added={}, changed={}, removed=removed)
+
+    def update(self, adict):
+        """ Update the values in the dict by the new dict.
+
+        Parameters
+        ----------
+        adict : dict
+            The dict from which values will be updated into this dict.
+        """
+
+        validated_dict = {}
+        added = {}
+        changed = {}
+
+        for key, value in adict.items():
+            validated_key = self.key_validator(key)
+            validated_value = self.value_validator(value)
+
+            if validated_key in self:
+                changed[validated_key] = self[validated_key]
+            else:
+                added[validated_key] = validated_value
+
+            validated_dict[validated_key] = validated_value
+
+        super().update(validated_dict)
+        self.notify(added=added, changed=changed, removed={})
+
+    def setdefault(self, key, value=None):
+        """ Returns the value if key is present in the dict, else creates the
+        key-value pair and returns the value.
+
+        Parameters
+        ----------
+        key : A hashable type.
+            Key to the item.
+        """
+
+        if key in self:
```

Comment:
This one is tricky because of the possible transformation that could happen on the key. e.g. the given key does not exist on the dict, but after the transformation, it exists. See test `test_setdefault_with_casting`.
If we want to call `super().setdefault` directly, then we would need to check the transformed key also does not exist in the dict so that it actually does `__setitem__`.

Reply:
The proposal in #1020 would allow this to be fixed cleanly. However, while I believe that proposal is the right way forward, the discussion on that proposal is postponed, so let's resolve this for now.